src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,340 @@
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async ValueTask PrepareBufferAsync(bool isAsync, CancellationToken ct)
+        {
+            // We have read all the data from the packet as stated in the header, this means that we have to 
+            // process the next packet header.
+            if (_packetDataLeft == 0 && _readBufferDataEnd > _readIndex)
+            {
+                await ProcessHeaderAsync(isAsync, ct).ConfigureAwait(false);
+            }
+
+            // There is no data left in the buffer.
+            if (_readIndex == _readBufferDataEnd)
+            {
+                // If we have left over data indicated in the packet header, then we simply need to get data from the network.
+                if (_packetDataLeft > 0)
+                {
+                    _readBufferDataEnd = isAsync ?
+                        await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                        _underlyingStream.Read(_readBuffer);
+                    _readIndex = 0;
+                }
+                // 1.2. There is no data left as indicated by packet header and the buffer is empty.
+                else if (_packetDataLeft == 0)
+                {
+                    _readBufferDataEnd = isAsync ?
+                        await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                        _underlyingStream.Read(_readBuffer);
+                    
+                    _readIndex = 0;
+
+                    await ProcessHeaderAsync(isAsync, ct).ConfigureAwait(false);
+
+                    // 1.3. After processing the packet header, there is a possibility that the transport read didn't
+                    // return any more data for the packet. In that case, post another read to have packet data ready..
+                    if (_readBufferDataEnd == _readIndex)
+                    {
+                        _readBufferDataEnd = isAsync ?
+                            await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                            _underlyingStream.Read(_readBuffer);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// If the buffer is empty, or the packet data is completely read, 
+        /// then we need to prepare the buffer for the next packet.
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async ValueTask PrepareBufferIfNeeded(bool isAsync, CancellationToken ct)
+        {
+            // If there we have read through the packet, 
+            // or if we have reached the end of the buffer,
+            // then we need to position the buffer at the beginning of packet or buffer
+            if (_packetDataLeft == 0 || _readBufferDataEnd == _readIndex)
```

Comment:
Nit: not sure if there's a rule about this, but there should be - even if it's one line, if statements should have curly braces